### PR TITLE
split seq2seq script into summarization & translation

### LINKED ISSUE #10164: [example scripts] disambiguate language specification API
Currently in example scripts like `run_seq2seq.py` we have:

1. for t5
```
--task translation_en_to_ro
--source_prefix "translate English to Romanian: "
```

2. Also these 2:
```
--target_lang ro_RO
--source_lang en_XX
```
are used only for MBart and are ignored for other models. Which means that people will unknowingly try to use these two as well when they aren't need.
 
The problem in both situations is that we provide error-prone API where a user wants to change the language and forgets that there is more than one of the same and changes only one of the sets of languages, but not the other, which leads to broken outcome.

If such an error is made the specification supplied by the user becomes ambiguous, because one can't tell which of the multiple inputs takes precedence.

Proposal: There should be only one way to input a set of languages and not multiple ways.

Specifically:

- in case 1, probably the easiest is to leave `--task translation_en_to_ro` and auto-generate `--source_prefix "translate English to Romanian: "`
- in case 2, assert if `--target_lang` or `--source_lang` are passed and the model is not MBart.

Thinking more about it, case 1 is a must to solve, because if a user misses `--source_prefix` or makes a typo in it - the train/eval won't fail, but will mysteriously produce really bad outcome. This is not user-friendly.

@sgugger, @patrickvonplaten, @patil-suraj

### PR DESCRIPTION
**Pull request by theo-m**
keeping the original script for tests cc @stas00 
fix #10164

### CODE REVIEW

File: docs/source/main_classes/trainer.rst
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:279]: Wondering if we should always use language codes that work for mBART (so en_XX and ro_RO) so the same command works well for those checkpoints.

File: docs/source/task_summary.rst
Comment (sgugger) [inline docs/source/task_summary.rst:826]:
```suggestion
translation task, you may leverage the `run_translation.py
<https://github.com/huggingface/transformers/tree/master/examples/seq2seq/run_translation.py>`__ script.
```

File: examples/seq2seq/README.md
Comment (sgugger) [inline examples/seq2seq/README.md:192]: T5 does not requires that, technically. It's required for the data preprocessing mostly.

File: examples/seq2seq/run_summarization.py
Comment (sgugger) [inline examples/seq2seq/run_summarization.py:352]: This part should be removed from this script as it's only relevant for translation.

File: examples/seq2seq/run_translation.py
Comment (sgugger) [inline examples/seq2seq/run_translation.py:60]: I don't think nltk is used in this script, so all of this can be removed.

File: examples/seq2seq/run_translation.py
Comment (sgugger) [inline examples/seq2seq/run_translation.py:276]: Adapt the comment since only JSON is supported :-)

File: examples/seq2seq/run_translation.py
Comment (sgugger) [inline examples/seq2seq/run_translation.py:220]: Only json is supported for translation so we should update the defensive tests.

**Review by sgugger — APPROVED** [review]
Thanks for working on this! It looks great already, I just left a few comments.

File: docs/source/installation.md
Comment (patil-suraj) [inline docs/source/installation.md:171]:
```suggestion
python examples/seq2seq/run_translation.py --model_name_or_path t5-small --dataset_name wmt16 --dataset_config ro-en ...
```

File: docs/source/installation.md
Comment (patil-suraj) [inline docs/source/installation.md:177]:
```suggestion
python examples/seq2seq/run_translation.py --model_name_or_path t5-small --dataset_name wmt16 --dataset_config ro-en ...
```

File: examples/seq2seq/README.md
Comment (patil-suraj) [inline examples/seq2seq/README.md:33]:
```suggestion
`run_summarization.py`  and  `run_translation.py` are lightweight examples of how to download and preprocess a dataset from the [🤗 Datasets](https://github.com/huggingface/datasets) library or use your own files (jsonlines or csv), then fine-tune one of the architectures above on it.
```

File: examples/seq2seq/run_summarization.py
Comment (patil-suraj) [inline examples/seq2seq/run_summarization.py:3]:
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

File: examples/seq2seq/run_summarization.py
Comment (patil-suraj) [reply examples/seq2seq/run_summarization.py]: `mbart-large-cc25` is a multilingual seq2seq model so can be used for summarization as well. I've seen few people using it for tasks other than translation.

File: examples/seq2seq/README.md
Comment (theo-m) [reply examples/seq2seq/README.md:192]: yup of course, but my current tradeoff (simplicity vs redundancy) makes them required (I raise a ValueError if these parameters aren't set), we can discuss the tradeoff but the doc is currently correct in this state.

File: docs/source/main_classes/trainer.rst
Comment (theo-m) [reply docs/source/main_classes/trainer.rst:279]: `en_XX` appeared quite mysterious so I favoured the simpler ISO codes since it appears that's what we do currently anyway: `.split("_")[0]`

**Review by patil-suraj — APPROVED** [review]
LGTM! Thanks a lot for working on this. Just left a few nits

File: examples/seq2seq/run_summarization.py
Comment (theo-m) [reply examples/seq2seq/run_summarization.py]: I've removed it in 68a8235, should I add it back?

File: examples/seq2seq/README.md
Comment (sgugger) [reply examples/seq2seq/README.md:192]: Yes, and I like that they are needed arguments. But with this change they are not "additional language-specific command-line arguments are sometimes required" as said above. I would remove the paragraph about mBART before, or say it requires custom source/target language codes like `en_XX` instead of `en` and just mention the `source_prefix` for T5 here.

File: docs/source/main_classes/trainer.rst
Comment (patil-suraj) [reply docs/source/main_classes/trainer.rst:279]: That might not work. `MBartTokenizer` needs such custom language codes, and they `src_lang` and `tgt_lang` are set before doing `.split("_")[0]`

File: docs/source/installation.md
Comment (theo-m) [reply docs/source/installation.md:171]: why would we prefer translation over summarization in this context?

**Comment by stas00** [comment]
> cc @stas00 `run_seq2seq` is left as is for now. At some point, if your tests migrate from that script to either the new ones or another one, we will remove it.

Please don't leave this on me - please sync the tests with these changes and remove the script. Thank you.

**Comment by stas00** [comment]
May I suggest that the examples are inconsistent with script naming-wise, some are very abbreviated `run_clm.py`, others are the extreme opposite `run_summarization.py` - that's a way too much to type - won't `run_sum.py` be sufficient?

File: docs/source/main_classes/trainer.rst
Comment (theo-m) [reply docs/source/main_classes/trainer.rst]: yes I've updated the doc as per [this comment](https://github.com/huggingface/transformers/pull/10611#discussion_r590554222), but I'd welcome comment w.r.t this: 
https://github.com/huggingface/transformers/blob/7df219b11607662c4ca9ba65988a296dff550d28/examples/seq2seq/README.md#L183

I have a feeling you might know where to find this doc?

**Comment by theo-m** [comment]
> May I suggest that the examples are inconsistent with script naming-wise, some are very abbreviated `run_clm.py`, others are the extreme opposite `run_summarization.py` - that's a way too much to type - won't `run_sum.py` be sufficient?

my personal preference goes to clarity so `run_summarization.py` trumps `run_sum.py` - but consistency is also important - unsure about the best tradeoff here

File: docs/source/installation.md
Comment (patil-suraj) [reply docs/source/installation.md:171]: `wmt16 ` is a translation dataset

**Comment by stas00** [comment]
`run_trans.py`

File: docs/source/main_classes/trainer.rst
Comment (patil-suraj) [reply docs/source/main_classes/trainer.rst]: The language codes can be found in the model card
https://huggingface.co/facebook/mbart-large-cc25

**Comment by stas00** [comment]
Also while we are at it or perhaps after the split - different PR, this further proposed improvement could be applied:
https://github.com/huggingface/transformers/issues/10164

File: examples/seq2seq/run_sum.py
Comment (stas00) [reply examples/seq2seq/run_sum.py]: It's a t5-only thing. needed for both tasks.

Comment (stas00) [reply examples/seq2seq/run_sum.py]: `input_ids = tokenizer("summarize: studies have shown that owning a dog is good for you "...`

Comment (stas00) [reply examples/seq2seq/run_sum.py]: That's why I cross-linked to https://github.com/huggingface/transformers/issues/10164 - currently the user has to spell out the source_prefix for translation, but not summarization.

**Comment by sgugger** [comment]
I'm not seeing the benefit of shortening to `run_sum` and `run_trans` when tab-complete will give you the full name. In the language-modeling folder, the acronym was used because `run_causal_language_modeling`/`run_masked_language_modeling` is really long, same for question-answering.

If summarization and translation took two words at least, we could use the acronym to shorten the name, but since they don't I would go for the full name.

**Comment by stas00** [comment]
> I'm not seeing the benefit of shortening to `run_sum` and `run_trans` when tab-complete will give you the full name. In the language-modeling folder, the acronym was used because `run_causal_language_modeling`/`run_masked_language_modeling` is really long, same for question-answering.
> 
> If summarization and translation took two words at least, we could use the acronym to shorten the name, but since they don't I would go for the full name.

I can't see how one is much longer than then other:
```
run_causal_language_modeling
run_summarization
```

Would `run_causal_lm` and `run_masked_lm` be perhaps a good middle ground?

**Comment by theo-m** [comment]
> I'm not seeing the benefit of shortening to `run_sum` and `run_trans` when tab-complete will give you the full name. In the language-modeling folder, the acronym was used because `run_causal_language_modeling`/`run_masked_language_modeling` is really long, same for question-answering.
> 
> If summarization and translation took two words at least, we could use the acronym to shorten the name, but since they don't I would go for the full name.

My current understanding is that consistency primes for now, and the change for more explicit script names should be done in a separate PR, if there is indeed consensus that explicit names are better here.

File: examples/seq2seq/README.md
Comment (stas00) [inline examples/seq2seq/README.md:33]:
```suggestion
`run_sum.py` and `run_trans.py` are lightweight examples of how to download and preprocess a dataset from the [🤗 Datasets](https://github.com/huggingface/datasets) library or use your own files (jsonlines or csv), then fine-tune one of the architectures above on it.
```

File: examples/seq2seq/README.md
Comment (stas00) [inline examples/seq2seq/README.md:170]: here where they are already there.

File: examples/seq2seq/README.md
Comment (stas00) [inline examples/seq2seq/README.md:167]:
```suggestion
```
these are already there, please scroll down a bit.

File: examples/seq2seq/README.md
Comment (stas00) [inline examples/seq2seq/README.md:183]:
```suggestion
* MBart models require different `--{source,target}_lang` values, e.g. in place of `en` it expects `en_XX`, for `ro` it expects `ro_RO`. The full MBart specification for language codes can be looked up [here](https://huggingface.co/facebook/mbart-large-cc25)
```

File: examples/seq2seq/run_sum.py
Comment (stas00) [inline examples/seq2seq/run_sum.py:258]: could you please check if the same is needed for mt5?

Perhaps a cleaner solution here would be to refer to the config object?

e.g. get the config object and then check `if config.architecture in [ "MT5ForConditionalGeneration", "T5WithLMHeadModel"]` I don't know if it's exhaustive.

I'm just looking at:

https://huggingface.co/t5-small/blob/main/config.json#L3
https://huggingface.co/google/mt5-small/blob/main/config.json#L3

**Review by stas00 — APPROVED** [review]
Awesome work, @theo-m! Thank you!

Left a few comments/questions/suggestions

File: examples/seq2seq/README.md
Comment (stas00) [reply examples/seq2seq/README.md:170]: hmm, weird I guess one can't delete lines in "suggestion" - I tried to delete those 2 lines.

File: examples/seq2seq/README.md
Comment (stas00) [reply examples/seq2seq/README.md:167]: oh, I see, it did work, it just ordered my comments incorrectly.

File: examples/seq2seq/run_trans.py
Comment (stas00) [inline examples/seq2seq/run_trans.py:228]: and so the instructions should be updated to remove the `--source_prefix` command then, right?

Or do we need to leave it for some custom user models?

Comment (theo-m) [reply examples/seq2seq/run_trans.py:228]: yes you're right I forgot to remove it from the docs

and yes currently I'm keeping the parameter and only do the automatic setting if it's not present, so users can override it - but I'm not confident this is the best: what do we favour in these scripts, simpler-tweakable code vs all-terrain-ready-to-use scripts? 
my understanding through this work is that we want the scripts to be the latter, hence my original choice.
if the former I'd remove the `source_prefix` arg altogether

File: examples/seq2seq/run_sum.py
Comment (patil-suraj) [reply examples/seq2seq/run_sum.py:258]: Regarding the T5 prefixes:

We had discussed this in a previous PR and I think @patrickvonplaten and I agreed that it's best to just leave it to the user to specify the correct prefix and avoid having such if/else logic. And add a note about that in the readme, so the user will know when and what prefix to pass. I'm a bit unsure why it's being added back.

Comment (theo-m) [reply examples/seq2seq/run_sum.py:258]: Yes that perspective is relevant to the tradeoff between simpler scripts vs ready-to-use scripts, as I outlined here: https://github.com/huggingface/transformers/pull/10611#discussion_r591169232

File: examples/seq2seq/run_trans.py
Comment (LysandreJik) [reply examples/seq2seq/run_trans.py:228]: Just to clarify further and to give a bit of context, the example scripts are made to be very simple and tweakable, rather than all-purpose. This has been the decision for a while now, and we do not accept requests to add features to example scripts anymore, as that would hinder understandability.

File: examples/seq2seq/run_sum.py
Comment (stas00) [reply examples/seq2seq/run_sum.py:258]: The problem is that it leads to:
1. To multiple points of language pair definition which leads to error-prone code. The language pair should be specified once
2. The software shouldn't subtly misbehave if some required flag is passed or not. At the very least the example script should require `--source_prefix` cl arg if the model is of type `*t5*` - it must not be optional.

Combining these 2 together is what @theo-m  implemented.

File: docs/source/installation.md
Comment (patrickvonplaten) [inline docs/source/installation.md:171]: I'm also very much in favor of `run_translation.py` and `run_summarization.py` - I don't see any benefit in shortening the name here. `trans` is a less known abbreviation than `clm` or `mlm` in NLP IMO. I also don't like `run_sum` , `sum` is often used for summarization. 

I'd rather change `mlm` to `masked_language_modeling` than use `run_sum` here

File: examples/seq2seq/run_sum.py
Comment (patrickvonplaten) [reply examples/seq2seq/run_sum.py:258]: I'm voting against this. 3 reasons from my side:

- This is not in line with "simple notebooks" that are easy to tweak. I think we somewhat decided that notebooks should not follow the one-size-fits-it-all approach.
- We add those lines only for 5 checkpoints, while we could very well just let the user define the prefix. A user that knows how T5 was trained can insert the prefix. Also, for a user that doesn't know T5 this happens automatically which could mean that she/he might not be aware afterward that a prefix has to be added to make summarization work.
- People that fine-tune a model like `t5-large` don't necessarily need the summarization prefix to work. `t5-large` can very well be fine-tuned without a prefix (not the strongest argument).

Comment (patrickvonplaten) [reply examples/seq2seq/run_sum.py:258]: This is also not scalable when we have other "multi-task" models that work for summarization. There is a very low chance that those need the same prefix.

Comment (patrickvonplaten) [reply examples/seq2seq/run_sum.py:258]: I think I would be fine with a warning if you guys think it's really important

File: examples/seq2seq/run_trans.py
Comment (patrickvonplaten) [reply examples/seq2seq/run_trans.py:228]: Agree with @LysandreJik here. This again is very T5 specific. Also we add the `lc` library only for T5 here - don't think it's worth adding a new dependency for just 5 checkpoints.

**Comment by patrickvonplaten** [comment]
Very nice PR overall! 

I don't really agree though with the `sum`, `trans` and the "one-size-fits-it-all" design choices. Maybe we can settle on displaying a warning for T5 @stas00 ?

**Comment by stas00** [comment]
> I don't really agree though with the `sum`, `trans` 

As long as other examples are consistently named then it works.

> and the "one-size-fits-it-all" design choices. 

It doesn't sound like we are reaching a consensus here. I outlined that entering the same data more than once in the same input is error-prone. Perhaps there is another way to fix this w/o "one-size-fits-it-all" 

> Maybe we can settle on displaying a warning for T5 @stas00 ?

Yes, please. At the very least.

**Comment by patrickvonplaten** [comment]
@theo-m, @stas00 - if it's fine for you maybe we can change the script names to `run_summarization.py` and `run_translation.py` then and replace the automatic setting of `prefix` for T5 with a warning instead. Would that work? I'm more than happy to merge this PR then

**Comment by stas00** [comment]
If the group prefers it this way then it is fine as you propose.

**Comment by patrickvonplaten** [comment]
> If the group prefers it this way then it is fine as you propose.

Okey, maybe @sgugger, @patil-suraj and @LysandreJik can give their final word then as well

**Comment by patil-suraj** [comment]
I would prefer explicit names  (`run_summarization.py` and `run_translation.py`) and not handling T5 prefixes automatically.

**Comment by sgugger** [comment]
I have already given my opinion on the names. For the `source_prefix` I have no strong opinion as I don't see any "good" solution sadly. I'm fine with the warning.

**Comment by theo-m** [comment]
Seems to me this could be merged, last call for maintainers @patrickvonplaten @sgugger @LysandreJik (thumbs up on this message will be interpreted as a go 😉 )

**Review by LysandreJik — APPROVED** [review]

**Pull request merged**
